hutch_python/happi.py

```diff
@@ -36,32 +50,62 @@ def get_happi_objs(db, hutch):
         A mapping from item name to item
     """
     # Load the happi Client
-    client = happi.Client(path=db)
-    containers = list()
-    if beamlines is None:
-        beamline_conf = []
+    if None not in (light_ctrl, lightpath):
+        client = light_ctrl.client
     else:
-        # Find upstream items based on lightpath configuration
-        beamline_conf = beamlines.get(hutch.upper())
-    # Something strange is happening if there are no upstream items
-    if not beamline_conf:
-        logger.warning("Unable to find lightpath for %s",
-                       hutch.upper())
-        beamline_conf = {}
-    # Add the complete hutch beamline
-    beamline_conf[hutch.upper()] = {}
-    # Base beamline
-    for beamline, conf in beamline_conf.items():
+        client = happi.Client(path=db)
+    containers = list()
+
+    if light_ctrl is None:
+        # lightpath was unavailable, search by beamline name
+        reqs = dict(beamline=endstation.upper(), active=True)
+        results = client.search(**reqs)
+        containers.extend(res.item for res in results)
+        return _load_devices(*containers)
+
+    # if lightpath exists, we can grab upstream devices
+    dev_names = set()
+    paths = light_ctrl.beamlines[endstation.upper()]
+    for path in paths:
+        dev_names.update(path)
+
+    # gather happi items for each of these
+    for name in dev_names:
+        results = client.search(name=name)
+        containers.extend(res.item for res in results)
+
+    # also any device with the same beamline name
+    # since lightpath only grabs lightpath-active devices
+    beamlines = set(it.beamline for it in containers)
+
+    for line in beamlines:
         # Assume we want hutch items that are active
-        reqs = dict(beamline=beamline, active=True)
+        # items can be lightpath-inactive
+        reqs = dict(beamline=line, active=True)
         results = client.search(**reqs)
-        blc = [res.item for res in results]
+        blc = [res.item for res in results
+               if res.item.name not in dev_names]
         # Add the beamline containers to the complete list
         if blc:
             containers.extend(blc)
         else:
             logger.warning("No items found in database for %s",
-                           beamline.upper())
+                           line.upper())
```

Comment:
One small snag, maybe?
```
WARNING  get_happi_objs - No items found in database for KFE
```

Reply:
I think there aren't any missing devices, I'm not sure why we hit this block

Reply:
Good catch.  I think the functionality is fine, but the error message is wrong.  In this case the LightController grabbed a set of devices with `beamline=KFE`, but must have gathered them all, resulting in the additional `blc` list being empty.

Reply:
I'll revise this error message